File: packages/cli/src/templates/cc-native/_cc-native/hooks/plan_questions_early.py
#!/usr/bin/env python3
"""UserPromptSubmit hook - injects Phase A clarification prompt in plan mode.

On the first prompt in plan mode (before any code exploration), injects
a system-reminder telling Claude to ask clarification questions via
AskUserQuestion before exploring the codebase.

Skips if questions were already asked this session.
"""

import json
import sys
from pathlib import Path

_hook_dir = Path(__file__).resolve().parent
_cc_native_lib_dir = _hook_dir.parent / "lib"
_shared_lib_dir = _hook_dir.parent.parent / "_shared" / "lib"
sys.path.insert(0, str(_cc_native_lib_dir))
sys.path.insert(0, str(_shared_lib_dir))

from utils import was_questions_asked
from base.hook_utils import load_hook_input
from base.logger import log_debug, log_info, log_warn, log_error


PHASE_A_PROMPT = """
## Plan Mode: Clarify Before Exploring

Use AskUserQuestion now — one call, 3-4 questions — before reading any code.

### Why This Matters
Once you explore the codebase, you anchor on what you find. Questions asked after exploration confirm your assumptions instead of challenging them. Ask now, while your interpretation is still flexible.

### What to Ask About
Only ask about things you cannot discover from code — the user's intent, constraints, history, and priorities:

- **Ambiguity:** If you can read this request two different ways, ask which interpretation is correct. Provide your top 2-3 readings as options.
- **Invisible context:** What does the user assume "everyone knows" about this system that isn't documented? What's obvious to them but hidden to you?
- **Success criteria:** What does "done well" look like beyond the literal request? What would make them rate this a 10?
- **Constraints and history:** Has this been attempted before? Are there parts of the system that are off-limits or sensitive?

### How to Select Questions
1. Generate 5+ candidate questions across the lenses above
2. For each, evaluate: "If they answered A vs B, would I explore different files or take a different approach?" If no — discard it.
3. Keep the 3-4 where different answers lead to meaningfully different exploration strategies
4. Frame each with 2-3 concrete options so the user can react rather than generate from scratch
""".strip()


def main() -> int:
    try:
        payload = load_hook_input()
        if not payload:
            return 0

        permission_mode = payload.get("permission_mode", "")
        if permission_mode != "plan":
            return 0

        session_id = str(payload.get("session_id", ""))
        if not session_id:
            log_debug("plan_questions_early", "No session_id, skipping")
            return 0

        # Get project root for context operations
        project_root = Path(payload.get("cwd", ".")).resolve()

        if was_questions_asked(session_id, project_root):
            log_debug("plan_questions_early", "Questions already asked, skipping")
            return 0

        log_info("plan_questions_early", "Plan mode detected, injecting Phase A prompt")
        print(f"<system-reminder>{PHASE_A_PROMPT}</system-reminder>")

    except Exception as e:
        from base.hook_utils import log_hook_error

        log_hook_error("plan_questions_early", e, "UserPromptSubmit")
        log_error("plan_questions_early", str(e))

    return 0


if __name__ == "__main__":
    from base.hook_utils import run_hook

    run_hook(main, "plan_questions_early")

File: packages/cli/src/templates/cc-native/_cc-native/hooks/plan_questions_early.ts
#!/usr/bin/env bun
/**
 * UserPromptSubmit hook — injects post-exploration clarification prompt in plan mode.
 *
 * After explore agents finish examining the codebase, injects a system-reminder
 * telling Claude to ask clarification questions via AskUserQuestion to narrow
 * the approach before drafting the plan.
 *
 * Skips if questions were already asked this session.
 */

import { getProjectRoot } from "../../_core/lib-ts/runtime/constants.js";
import { loadHookInput, runHook, logDebug, logInfo, emitContext } from "../../_core/lib-ts/hooks/hook-utils.js";
import { wasEarlyQuestionsAsked } from "../lib-ts/cc-native-state.js";

// Unconditional injection by design — no code-detection gate.
// "When this plan involves code" is self-selecting; non-code plans ignore it.
// Soft framing per Anthropic Claude 4.x best practices (avoid MUST/MANDATORY overtriggering).
// Motivation per standard enables generalization better than threats.
// Generalizability disclaimer: not all codebases need all standards.
const CODING_STANDARDS_NUDGE = `## Coding Standards for Code Changes

When this plan creates or modifies production code, apply these standards — they address the
most common plan review failure modes:

1. **Test-First Design** — Design interfaces from the test perspective first. Plans that
   describe "implement then test" consistently fail review. Structure tests before implementation.
2. **File Structure Fit** — Verify where similar things already live in this project before
   proposing new files. Agents commonly pick plausible-but-wrong locations that don't match
   existing conventions.
3. **Extensibility Analysis** — Identify what features most commonly follow this one. Designs
   that resist extension require expensive rewrites later.

These standards apply to production code in established codebases. For prototypes, scripts,
or exploratory work, use judgment on which apply.

**Full checklist:** \`.aiwcli/_cc-native/plan-review/CODING-STANDARDS-CHECKLIST.md\`
Read this file for detailed guidance on each standard.`;

const PHASE_A_PROMPT = `## Plan Mode: Narrow the Approach After Exploration

After exploring the codebase, use AskUserQuestion — one call, 3-4 questions — before drafting the plan.

### Why This Matters
Once you've explored the codebase, you'll understand what exists — but not which direction the user prefers. That's a branch point: multiple viable approaches, and the user's priorities determine which is best. Questions asked after exploration have maximum steering value: they narrow your path before you commit to an implementation direction.

### What to Ask About
Only ask about decisions that exploration will surface but can't resolve — where human judgment is needed to choose between viable options:

- **Approach selection:** If exploration reveals 2-3 viable implementation paths, ask which the user prefers. Present each option with its trade-offs as concrete choices.
- **Scope boundaries:** What's in scope vs. out of scope for this change? Which areas of the codebase should be left untouched? How far should the change ripple?
- **Trade-off preferences:** Where exploration reveals tensions (simplicity vs. flexibility, speed vs. thoroughness, minimal change vs. full refactor), ask which side the user leans toward.
- **Success criteria beyond the literal ask:** What would make this a 10? What non-obvious quality matters most — performance, readability, extensibility, consistency with existing patterns?

### How to Select Questions
1. Generate 5+ candidate questions across the categories above
2. For each, evaluate: "If they answered A vs B, would I take a different approach or write different code?" If no — discard it.
3. Keep the 3-4 where different answers lead to meaningfully different implementation strategies
4. Frame each with 2-3 concrete options so the user can react rather than generate from scratch`;

function main(): void {
  const payload = loadHookInput();
  if (!payload) return;

  const permissionMode = payload.permission_mode ?? "";
  if (permissionMode !== "plan") return;

  const sessionId = String(payload.session_id ?? "");
  if (!sessionId) {
    logDebug("plan_questions_early", "No session_id, skipping");
    return;
  }

  const projectRoot = getProjectRoot(payload.cwd);

  if (wasEarlyQuestionsAsked(sessionId, projectRoot)) {
    logDebug("plan_questions_early", "Early questions already asked, skipping Phase A prompt");
    return;
  }

  logInfo("plan_questions_early", "Plan mode detected, injecting Phase A prompt");
  emitContext(PHASE_A_PROMPT);
  emitContext(CODING_STANDARDS_NUDGE);
}

runHook(main, "plan_questions_early");
